zilliqa-macros/src/test.rs

```diff
@@ -22,33 +22,45 @@ pub(crate) fn test_macro(_args: TokenStream, item: TokenStream) -> TokenStream {
             // The original test function
             #input
 
+            let run_sequentially = std::env::var_os("ZQ_RUN_SEQUENTIALLY").is_some();
+
             // Work out what RNG seeds to run the test with.
             let seeds: Vec<u64> = if let Some(seed) = std::env::var_os("ZQ_TEST_RNG_SEED") {
                 vec![seed.to_str().unwrap().parse().unwrap()]
             } else {
                 let samples: usize = std::env::var_os("ZQ_TEST_SAMPLES")
                     .map(|s| s.to_str().unwrap().parse().expect(&format!("Failed to parse ZQ_TEST_SAMPLES env var: {:?}", s)))
                     .unwrap_or(1);
-                // Generate random seeds using the thread-local RNG.
-                rand::Rng::sample_iter(rand::thread_rng(), rand::distributions::Standard).take(samples).collect()
+                (0..samples).map(|x| x as u64).collect()
```

Comment:
That is correct. I think it is cleaner to have it like this because:
A) When you look at the seed it failed on you get an idea of how many seeds it took until failure.
B) You are guaranteed that you are running the same seeds as the CI.

The slight issue is that the way it is currently, the CI will spit out that you failed and replicate this test by running seed X.

However, since all of the tests are running async, I believe it is the case that the last seed printed out doesn't neccessarily correspond to the failing test. This makes it a bit easier to replicate the failing CI env locally.

The downside is of course that the tests will be run against seeds 0-100 forever so there is some loss of coverage as it were.

Reply:
I'm not a big fan, mainly because I'm worried about the loss of coverage. Now in theory the CI should have enough samples to uncover failures in any case, but given that we're randomising the messages anyway this will always be probabilistic and therefore a tradeoff. If there's some message ordering that just happens to never come up on the seeds 0-100, it's guaranteed to never come up in CI now, whereas currently any given CI run _might_ miss an error but would then likely catch whenever it gets re-ran (which means very soon, considering CI tends to get ran multiple times per PR, and we open new PRs regularly).

Locally too, when you're running a single sample (or sometimes I run with say 3-5, but rarely more), this makes it more annoying because mid-development you're likely to "overfit" to seed 0 and if there's any consensus bugs, you either have to run the full CI test, or manually randomise the seed, or you won't discover them until you push and wait for CI. Whereas currently ever `cargo test` is randomised so you gain confidence that your changes are probably mostly robust when tests pass.

> The slight issue is that the way it is currently, the CI will spit out that you failed and replicate this test by running seed X.
> However, since all of the tests are running async, I believe it is the case that the last seed printed out doesn't neccessarily correspond to the failing test.

Does this change actually improve that, though? If the latest number printed doesn't necessarily correspond to the failure, I don't see how making the numbers sequential would necessarily fix things.
And overall, while I agree that it's slightly annoying to isolate a failing seed, it usually only takes me a couple of minutes to bisect the likely candidates and find the one(s) responsible, and I'd rather keep the randomised coverage than make this step slightly easier at the expense of only ever testing the same 100 seeds henceforth.

Reply:
I agree with @theo-zil here.

Right now, we explore a different area of the large search space of possible computations every time anything runs the test suite. In my opinion, this can give us a high degree of confidence that our code is correct under most conditions.

If we wanted to test the same set of computations every time, I would argue we need to dramatically increase the number of samples we test, because we don't have the benefit of every subsequent PR contributing to further tests of previous merges.

However, I think #534 might address some the issues you mention in a different way @n-hutton?

Reply:
Okay - with James' PR it makes it much more usable so I will revert this